### PR TITLE
ref(notifications): Move email code out of notifications module

### PR DESCRIPTION
**Pull request by mgaeta**
This refactor cleans up /src/notifications/activity/base.py in three ways:
- Clean up `context` inheritance
- Move Slack/email-specific code to their own files
- Update the notification functions to take a list of users and move the registry to a file

### CODE REVIEW

File: tests/sentry/integrations/slack/test_notifications.py [hunk 1]
```diff
@@ -50,6 +60,12 @@ def setUp(self):
             NotificationSettingOptionValues.ALWAYS,
             user=self.user,
         )
+        NotificationSetting.objects.update_settings(
```
Comment (ceorourke) [inline tests/sentry/integrations/slack/test_notifications.py:63]: I suppose this doesn't hurt to have but there aren't any tests targeting issue alerts in this file

File: src/sentry/mail/activity.py [hunk 2]
```diff
@@ -1,24 +1,95 @@
...
+
+    return context
+
+
+@register_notification_provider(ExternalProviders.EMAIL)
 def send_notification_as_email(
-    notification: ActivityNotification, user: User, context: Mapping[str, Any]
+    notification: ActivityNotification,
+    users: Mapping[User, int],
+    shared_context: Mapping[str, Any],
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/mail/activity.py:77]: `user_context` is per-user and `shared_context` is per-notification or "shared" by all users.

File: src/sentry/mail/activity.py [hunk 3]
```diff
@@ -1,24 +1,95 @@
...
-        html_template=notification.get_html_template(),
-        headers=notification.get_headers(),
-        type=notification.get_email_type(),
-        context=context,
-        reference=notification.activity,
-        reply_reference=notification.group,
-    )
-    msg.add_users([user.id], project=notification.project)
-    msg.send_async()
+    headers = get_headers(notification)
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/mail/activity.py:79]: I wanted to make the `MessageBuilder` a little cleaner.

File: tests/sentry/integrations/slack/test_notifications.py (hunk 1 above)
Comment (mgaeta) [reply tests/sentry/integrations/slack/test_notifications.py:63]: `NewProcessingIssuesActivityNotification` reads from "issue alert" notification settings.

File: src/sentry/mail/activity.py [hunk 4]
```diff
@@ -1,24 +1,95 @@
...
+def get_context(
+    notification, user: User, reason: int, shared_context: Mapping[str, Any]
+) -> Mapping[str, Any]:
+    """
+    Compose the various levels of context and add email-specific fields. The
+    generic HTML/text templates only render the unsubscribe link if one is
+    present in the context, so don't automatically add it to every message.
+    """
+    context = {
+        **shared_context,
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/mail/activity.py:64]: Yes they're in `.get_context()`.

File: tests/sentry/integrations/slack/test_notifications.py (hunk 1 above)
Comment (ceorourke) [reply tests/sentry/integrations/slack/test_notifications.py:63]: hmmm then I'm confused how the test passed without it, maybe it was default before.

**Review by ceorourke — APPROVED** [review]

**Pull request merged**